File: reconstruct-original-digits-from-english/reconstruct-original-digits-from-english.cpp
class Solution {
public:
    string originalDigits(string s) {
        vector<char> filter_order{'z','w','u','x','g','o','h','f','s','i'};
        vector<int> filter_index{0,2,4,6,8,1,3,5,7,9};
        vector<string> num{"zero","one","two","three","four","five","six","seven","eight","nine"};
        vector<int> letter_freq(26,0);
        for(int i = 0 ; i < s.size() ; i++){
            letter_freq[s[i] - 'a']++;
        }
        string ans = "";
        for(int i = 0 ; i < filter_order.size() ; i++){
            if(letter_freq[filter_order[i] - 'a']){
                int sub = letter_freq[filter_order[i] - 'a'];
                for(int p = 0 ; p < sub ; p++){
                    ans+=('0'+filter_index[i]);
                }
                for(int j = 0 ; j < num[filter_index[i]].size() ; j++){
                    letter_freq[num[filter_index[i]][j] - 'a']-=sub;
                }
            }
        }
        sort(ans.begin(),ans.end());
        return ans; 
    }
};

File: reconstruct-original-digits-from-english/reconstruct-original-digits-from-english.go
import "sort"
import "strconv"
type sortRune []rune

func (s sortRune) Less(i,j int) bool{
    return s[i] < s[j]
}

func (s sortRune) Len()int{
    return len(s)
}

func (s sortRune) Swap(i,j int) {
    s[i],s[j] = s[j],s[i]
}



func originalDigits(s string) string {
    filter_order := []rune{'z','w','u','x','g','o','h','f','s','i'}
    filter_index := []int{0,2,4,6,8,1,3,5,7,9}
    num          := []string{"zero","one","two","three","four","five","six","seven","eight","nine"}
    letter_freq  := make([]int,26)
    for i := range letter_freq {
        letter_freq[i] = 0
    }
    
    for  i := 0 ; i < len(s) ; i++ {
        letter_freq[int(s[i]) - int('a')]+=1
    }
    ans := ""
    for i := 0 ; i < len(filter_order) ; i++ {
        if letter_freq[int(filter_order[i]) - int('a')] != 0 {
            sub := letter_freq[int(filter_order[i]) - int('a')]
            for p := 0 ; p < sub ; p++ {
                ans += strconv.Itoa(filter_index[i])
            }
            for j := 0 ; j < len(num[filter_index[i]]) ; j++ {
                letter_freq[int(num[filter_index[i]][j]) - int('a')]-=sub
            }
        }
    }
    ansRune := make(sortRune,0)
    for _,runeValue := range ans {
        ansRune = append(ansRune , runeValue)
    }
    sort.Sort(ansRune)
    answer := string(ansRune)
    return answer
}